heartbeat/openstack-virtual-ip

```diff
@@ -0,0 +1,275 @@
+</parameter>
+
+<parameter name="subnet_id" required="1">
+<longdesc lang="en">
+Subnet Identifier to use to attach the address.
+</longdesc>
+<shortdesc lang="en">Subnet ID</shortdesc>
+<content type="string" />
+</parameter>
+
+</parameters>
+
+<actions>
+<action name="start" timeout="180s" />
+<action name="stop" timeout="180s" />
+<action name="monitor" depth="0" timeout="30s" interval="60s" />
+<action name="validate-all" timeout="5s" />
+<action name="meta-data" timeout="5s" />
+</actions>
+</resource-agent>
+END
+}
+
+osvip_port_id() {
+	# Get port_id from subnet_id
+	node_port_ids=$(${HA_SBIN_DIR}/attrd_updater --query -n openstack_ports -N $(crm_node -n) \
+		| awk '{gsub("value=","") ; gsub("\"","") ; print $NF}')
+
+	node_port_id=$(echo $node_port_ids \
+		| tr ',' '\n' \
+		| awk -F: "/$OCF_RESKEY_subnet_id/ {print \$2}")
+
+	echo ${node_port_id}
+}
+
+osvip_validate() {
+	check_binary "$OCF_RESKEY_openstackcli"
+
+	if [ -z "$OCF_RESKEY_openrc" ]; then
+		ocf_exit_reason "openrc parameter not set"
+		return $OCF_ERR_CONFIGURED
+	fi
+
+	if [ ! -f "$OCF_RESKEY_openrc" ] ; then
+		ocf_exit_reason "openrc file not found"
+		return $OCF_ERR_CONFIGURED
+	fi
+
+	. $OCF_RESKEY_openrc
+
+	${HA_SBIN_DIR}/attrd_updater --query -n openstack_ports -N $(crm_node -n) > /dev/null 2>&1
+	if [ $? -ne 0 ] ; then
+		ocf_log warn "attr_updater failed to get openstack_ports attribute of node $OCF_RESOURCE_INSTANCE"
+		return $OCF_ERR_GENERIC
+	fi
+
+	return $OCF_SUCCESS
+}
+
+osvip_monitor() {
+	local result
+
+	node_port_id=$(osvip_port_id)
+
+	result=$($OCF_RESKEY_openstackcli port show \
+		--format value \
+		--column allowed_address_pairs \
+		${node_port_id})
+	if echo $result | grep -q $OCF_RESKEY_ip ; then
+		${HA_SBIN_DIR}/attrd_updater ${OCF_RESKEY_delay} -S status -n openstack_virtual_ip -v $OCF_RESKEY_ip
+
+		return $OCF_SUCCESS
+	fi
+
+	${HA_SBIN_DIR}/attrd_updater ${OCF_RESKEY_delay} -D -S state -n openstack_virtual_ip
+	ocf_log warn "$OCF_RESKEY_ip is not attached to any fixed address"
+	return $OCF_NOT_RUNNING
+}
+
+osvip_stop() {
+	node_port_id=$(osvip_port_id)
+
+	ocf_log info "Bringing down IP address $OCF_RESKEY_ip"
+
+	osvip_monitor
+	if [ $? = $OCF_NOT_RUNNING ]; then
+		ocf_log info "Address $OCF_RESKEY_ip already down"
+		return $OCF_SUCCESS
+	fi
+
+	mac_address=$($OCF_RESKEY_openstackcli port show \
+		--format value \
+		--column mac_address \
+		$node_port_id)
+	echo ${mac_address} | grep -q -P "^([0-9a-f]{2}:){5}[0-9a-f]{2}$"
+	if [ $? -ne 0 ]; then
+		ocf_log error "MAC address '${mac_address}' is not valid."
+		return $OCF_ERR_GENERIC
+	fi
+
+	if ! $OCF_RESKEY_openstackcli port unset \
+		--allowed-address \
+		ip-address=$OCF_RESKEY_ip,mac-address=${mac_address} \
+		$node_port_id; then
+		return $OCF_ERR_GENERIC
+	fi
+
+	osvip_monitor
+	if [ $? != $OCF_NOT_RUNNING ]; then
+		ocf_log error "Couldn't unset IP address $OCF_RESKEY_ip."
+		return $OCF_ERR_GENERIC
+	fi
+
+	ocf_log info "Successfully brought unset $OCF_RESKEY_ip"
```

Comment:
This sounds a bit strange to me, and I think detached is a better word for unset in the situation.

Reply:
Fixed with commit 61aebd2c482b0052b2865fba01acc11bc0fc5fab:
```
-	ocf_log info "Successfully brought unset $OCF_RESKEY_ip"
+	ocf_log info "Successfully brought down $OCF_RESKEY_ip"
```